lib/workflow/messengers/messenger-AMQP.js

```diff
@@ -20,67 +21,121 @@ function amqpMessengerFactory(
 ) {
-    var exports = {};
+    var logger = Logger.initialize(amqpMessengerFactory);
 
-    exports.subscribeRunTask = function(domain, callback) {
+    function AMQPMessenger() {
+    }
+
+    AMQPMessenger.prototype.subscribeRunTask = function(domain, callback) {
         return taskProtocol.subscribeRun(domain, callback);
     };
 
-    exports.publishRunTask = function(domain, taskId, graphId) {
+    AMQPMessenger.prototype.publishRunTask = function(domain, taskId, graphId) {
         return taskProtocol.run(domain, { taskId: taskId, graphId: graphId });
     };
 
-    exports.subscribeCancelTask = function(callback) {
+    AMQPMessenger.prototype.subscribeCancelTask = function(callback) {
         return taskProtocol.subscribeCancel(callback);
     };
 
-    exports.publishCancelTask = function(taskId, errName, errMessage) {
+    AMQPMessenger.prototype.publishCancelTask = function(taskId, errName, errMessage) {
         return taskProtocol.cancel(taskId, errName, errMessage);
     };
 
-    exports.subscribeTaskFinished = function(domain, callback) {
+    AMQPMessenger.prototype.subscribeTaskFinished = function(domain, callback) {
         return eventsProtocol.subscribeTaskFinished(domain, callback);
     };
 
-    exports.publishTaskFinished = function(
-        domain,
-        taskId,
-        graphId,
-        state,
-        error,
-        context,
-        terminalOnStates
-    ) {
+    /**
+     * Publishes a task finished event over AMQP
+     *
+     * @param {String} domain
+     * @param {Object} task
+     * @returns {Promise}
+     * @memberOf AMQPMessenger
+     */
+    AMQPMessenger.prototype.publishTaskFinished = function(domain, task) {
+        var errorMsg;
+        if (task.error && task.error.stack) {
+            errorMsg = task.error.stack;
+        } else if (task.error) {
+            errorMsg = task.error.toString();
+        }
         return eventsProtocol.publishTaskFinished(
-                domain,
-                taskId,
-                graphId,
-                state,
-                error,
-                context,
-                terminalOnStates
-        );
+            domain,
+            task.instanceId,
+            task.context.graphId,
+            task.state,
+            errorMsg,
+            task.context,
+            task.definition.terminalOnStates
+        )
+        .catch(function(error) {
+            logger.error('Error publishing task finished event', {
+                taskId: task.instanceId,
+                graphId: task.context.graphId,
+                state: task.state,
+                error: error
+            });
+        });
+    };
+
+    /**
+     * Publish a graph started event over AMQP.
+     *
+     * @param {Object} graph
+     * @returns {Promise}
+     * @memberOf AMQPMessenger
+     */
+    AMQPMessenger.prototype.publishGraphStarted = function(graph) {
+        return eventsProtocol.publishGraphStarted(graph.instanceId, graph._status, graph.node)
```

Comment:
My original purpose is that workflow related events only rely on `Task.Messenger` instead of both `Protocol.Events` and `Task.Messenger`. But I will move `eventsProtocol.publishGraphStarted` out of AMQPMessenger since it also publishes message via Web Hooks.